Null values in html attribute lists trigger unexpected exceptions

gsteel opened the pull request:
|    Q          |   A
|-------------- | ------
| Bugfix        | yes
| QA            | yes

Prior to 2.13.x, providing an array of attributes such as `['class' => 'whatever', 'data-something' => null]` to a helper that extends from `AbstractHtmlElement` would yield markup along the lines of `class="whatever" data-something=""`

Now, a type error is thrown in `HtmlAttributesSet::__toString()` causing a fatal error because of the the throw in __toString

gsteel committed:
Extend abstract element with a concrete helper to expose protected methods and add a failing tests for null attribute values

Signed-off-by: George Steel <george@net-glue.co.uk>

froschdesign commented:
> Prior to 2.13.x, providing an array of attributes such as `['class' => 'whatever', 'data-something' => null]` to a helper that extends from `AbstractHtmlElement` would yield markup along the lines of `class="whatever" data-something=""`

This was never tested either. :sob:

gsteel commented:
@froschdesign - I'm just working on fixing the HtmlAttributesSet class and providing tests for it. There will be BC breaks to HtmlAttributesSet, but considering it's been in the wild for less than a day, I think those breaks should be acceptable.

gsteel committed:
Fix mistakes in initial failing tests

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Fix issues with null attributes and other problems…

- `HtmlAttributesSet` uses json_encode so ext-json is declared in require. This also updates the lock file
- Simplifies the constructor to HtmlAttributesSet. Only 1 escaper instance is required.
- Fixes type annotations for iterables in HtmlAttributesSet
- Casts scalars to strings during escaping to avoid type errors with the escaper.
- Adds an initial unit test to cover HtmlAttributesSet
- Adds additional tests to `AbstractHtmlElement` to ensure attributes are escaped correctly

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
CS Fix

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel commented:
Hey @froschdesign, please could you review?

gsteel commented on src/Helper/HtmlAttributes.php:
```diff
@@ -18,12 +19,20 @@ class HtmlAttributes extends AbstractHelper
+        assert($escapePlugin instanceof AbstractEscapeHelper);
+        $escaper = $escapePlugin->getEscaper();
+        assert($escaper instanceof Escaper);
+
```
All those assertions keep psalm happy. Could we not just instantiate our own Escaper in __construct() and avoid all of that? i.e. `public function __construct(Escaper $escaper = null) { $this->escaper = $escaper ?? new Escaper(); }`

froschdesign replied:
But this means that if the escaper is to be replaced, I have to add / write a factory for this helper too.
That would not be enough

```php
'factories' => [
    Helper\EscapeHtml::class     => MyEscaperFactory::class,
    Helper\EscapeHtmlAttr::class => MyEscaperFactory::class,
    Helper\EscapeJs::class       => MyEscaperFactory::class,
    Helper\EscapeCss::class      => MyEscaperFactory::class,
    Helper\EscapeUrl::class      => MyEscaperFactory::class,
]
```
(And that in itself is _unbeautiful_.)

gsteel replied:
👍 - I figured that others might balk at all the assertions.

gsteel commented on src/HtmlAttributesSet.php:
```diff
@@ -21,40 +29,26 @@ class HtmlAttributesSet extends ArrayObject
-     * @param iterable $attributes Attributes to manage
-     */
-    public function __construct(Escaper $htmlEscaper, Escaper $htmlAttributeEscaper, iterable $attributes = [])
+    public function __construct(Escaper $escaper, iterable $attributes = [])
```
How many escapers do we actually need?

froschdesign replied:
One, because there is only one class which offers all methods for HTML, HTML attributes, Javascript, CSS and URL.

https://docs.laminas.dev/laminas-escaper/intro/#overview

gsteel replied:
Sorry, I was attempting a little bit of sarcasm there!

froschdesign replied:
And I gave you an _expert answer_, so it worked! 😄

gsteel commented:
Good morning @froschdesign - Would it be possible to get this into a patch release today? I've got projects queued up for deploy that are failing CI due to this bug. Sorry to nag 😬

gsteel committed:
This class should be marked as final whilst we're breaking BC with it anyway

Signed-off-by: George Steel <george@net-glue.co.uk>

froschdesign commented:
@gsteel 
_Sorry for the delay. I have a lot of client projects at the moment._

Thank you for your time and this contribution! 👍
